catalog: fix descriptor validation early exit

postamar opened the pull request:
This patch fixes a bug in doctor and crdb_internal.invalid_objects
preventing multiple validation errors originating from different
validation levels from being properly reported.

Release note: None

Commit:
catalog: fix descriptor validation early exit

This patch fixes a bug in doctor and crdb_internal.invalid_objects
preventing multiple validation errors originating from different
validation levels from being properly reported.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62878)
<!-- Reviewable:end -->

postamar commented:
This bug was introduced by https://github.com/cockroachdb/cockroach/pull/62546 and I stumbled upon it while fixing a merge conflict for https://github.com/cockroachdb/cockroach/pull/62803 . I might otherwise never have as it's quite an edge case so I'm not too sad. I added a doctor test case to cover this.

Commit:
sql: improve telemetry test cases for descriptor corruption

This patch improves the deliberate corruption of the descriptor state in
the `error` suite of the telemetry logic tests.

Release note: None

postamar commented:
In the same vein of improving the telemetry PR I've improved the test cases which I was having trouble backporting to 21.1. These are now more robust in any case.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62878#-:-MXFTT7H6wW49PeulyHW:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->

postamar commented:
bors r+
